### PR TITLE
[ci] Fix BIN_DIR merging and artifact passing between jobs

### LINKED ISSUE #5029: spiflash on FPGA is timing out
On master, the FPGA tests currently time out because they are unable to load the application image to the FPGA using the spiflash utility. This was first seen on master after https://github.com/lowRISC/opentitan/pull/5002 got merged (which passed in PR CI). Since this PR got merged, all master builds consistently failed.

Since the PR check passed, it's likely a combination of two commits that actually caused the issue. 

Full log (not very helpful): https://dev.azure.com/lowrisc/4c50229e-f6cb-4f39-8522-585bb3841438/_apis/build/builds/32144/logs/194

--- 

Update 2021-02-01 14:30 GMT:
- In CI we now ignore failures from FPGA tests (#5043), allowing us to resume merging into master.
- The root cause of the issue is as unclear as before, @vogelpi and me are looking into it.

### LINKED ISSUE #5044: Package nexysvideo build artifacts into dist archive which work
We currently build the SW twice, once "for real", and once for the Nexys Video board with reduced flash sizes. We currently always pack the "real" software into the distribution archive, giving us files with "nexysvideo" in the name, which are actually not runnable there (because they are not size-reduced). We need to instead patch the distribution archive together from two software builds, taking the files with nexysvideo in the name from the nexysvideo-specific build.

### PR DESCRIPTION
**Pull request by imphil**
The CI build pipelines is split into different jobs. Build artifacts between jobs are passed in the `BIN_DIR` directory. Due to a variety of misunderstandings and bugs, this passing wasn't working as expected:

* Multiple jobs produced identical artifacts in BIN_DIR, i.e. we didn't have an unambiguous build rule for each output in BIN_DIR any more.
* Partial BIN_DIR artifacts were downloaded from unrelated jobs in a non-determinisic manner.
* The overall system of selecting intended build artifacts and merging partial BIN_DIR outputs into a final result wasn't very robust.

This series of commits fixes the underlying problems and makes the whole build pipeline a bit more robust and explicit; see the individual commit messages for details.

In the medium term, much of the logic contained in the pipeline definitions today should be part of a higher-level build system, which, for example, takes care of having unambiguous ways to produce output artifacts, and ensures that the right input artifact dependencies are present or built if needed. That work has started and will hopefully come to fruition soon. In the meantime, I went for a fix which improves on the status quo without going overboard.

Thanks a lot to @vogelpi for ultimately finding the reason why builds were failing when I got lost in the weeds of synthesis logs.

Fixes #5029
Fixes #5044

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -456,6 +472,13 @@ jobs:
   steps:
   - template: ci/install-package-dependencies.yml
   - template: ci/download-artifacts-template.yml
+    parameters:
+      downloadPartialBuildBinFrom:
+        - sw_build
+        - sw_build_nexysvideo
+        - top_earlgrey_verilator
+        - top_earlgrey_nexysvideo
+        - top_englishbreakfast_verilator
```
[the remark this answers is not in the document]
Comment (vogelpi) [reply azure-pipelines.yml:500]: Ah I see this is clarified in the second to last commit message. This commit also removes the override flag. So, all good.

**Review by vogelpi — APPROVED** [review]
Many thanks for investigating and solving this critical issue @imphil !

File: azure-pipelines.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (imphil) [reply azure-pipelines.yml:500]: Exactly: The order does not matter, the sets of files within those partial $BIN_DIRs must be disjoint, and `tar` enforces that.

**Review by mcy — APPROVED** [review]
Looks good. Hopefully the situation will improve sooner rather than later. =/

**Comment by imphil** [comment]
Thanks for your reviews, I've updated the comments a bit and merged the result.

**Comment by tjaychen** [comment]
thanks @imphil and @vogelpi for fixing this!

**Pull request merged**